### ISSUE TITLE
online vs offline triplets mining? hard, semihard, easy mining?

### ISSUE THREAD

**Issue by KennyTC** [KevinMusgrave/pytorch-metric-learning#450]
I have read through the definitions of online and offline triplet mining, but not quite understand it. 
As far as I can understand, 
-  offline mining: we calculate the embeddings of all instances in train data, then select only hard/semihard negatives triplets in advance (suppose we have totally N semihard/hard triplets). During training, if the batch size=m, then each time, we will choose m out of these  N triplets.
- online mining:  we do not select hard/semi-hard negatives triplets in advance,  but we will select them during training?  (i.e, we will load m instances into memory, and based on its labels, we will calculate hard/semi-hard negatives within this batch)
1) Does the above is correct?  If yes, then what is the benefit of online mining over offline mining?
2) Is current implementation of this library
https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/miners/triplet_margin_miner.py 
called online mining? 
3) As far as I know, we should avoid "easy triplets" or "easy negatives". But in triplet_margin_miner.py, we have  options type_of_triplets＝"easy", what is the benefit of choosing "easy triplets"?
4) I'm not quite sure but it seems that "hard negatives" not always better than "semihard negatives". Do you know why?

I know  that many of the above questions may not relate directly to your library, but triplet loss-related general 
 questions. But since you are expert on this field, I hope that you can help me to clarify.

Thanks a lot for your amazing library.

**Comment by KevinMusgrave** [comment]
> 1. Does the above is correct?  If yes, then what is the benefit of online mining over offline mining

Yes that's correct. I think the main benefit of online mining is that it always uses the current state of the model, rather than an old state. When optimizing the model, it's more useful to know what it finds difficult at the current iteration, rather than what it found difficult N iterations ago.

> 2. Is current implementation of this library
>    https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/miners/triplet_margin_miner.py
>    called online mining?

Yes

> 3. As far as I know, we should avoid "easy triplets" or "easy negatives". But in triplet_margin_miner.py, we have  options type_of_triplets＝"easy", what is the benefit of choosing "easy triplets"?

There's no benefit that I know of. The option is there just in case someone wants to access easy triplets.

> 4. I'm not quite sure but it seems that "hard negatives" not always better than "semihard negatives". Do you know why?

[This old paper](https://arxiv.org/pdf/1503.03832.pdf) says: "Selecting the hardest negatives can in practice lead to bad local minima early on in training, specifically it can result in a collapsed model". So that is one possible explanation.

**Comment by KennyTC** [comment]
It is clear to me. Thanks very much for your helpful library.
